File: vscode-extension/src/parser/scriptUndo.js
// Port of core/script/scriptUndo.cpp + scriptHistory.cpp — the per-block edit ledger and
// the `@undo` / `@redo` statements that drive it.
import { makeDiag, tokenOfStmt } from './scriptDiagnostics.js';

/* Edits are numbered 1..n as written and never renumbered. Undoing marks an edit dead; the
 * ledger then RE-DERIVES the image state at each @save by emitting one undo{steps} back to
 * the last agreeing edit and replaying the survivors. No adapter computes an undo count. */
export class EditLedger {
  #edits = [];

  #applied = [];

  #removed = []; // LIFO, what @redo brings back

  addEdit(op, text) {
    this.#edits.push({ op, text, live: true });
    this.#applied.push(this.#edits.length - 1);
    return this.#edits.length;
  }

  get editCount() {
    return this.#edits.length;
  }

  // Selector forms: 0 = the last live edit, N = that edit, -N = N-th from the end.
  undoIndex(selector) {
    const n = this.#edits.length;
    if (n === 0) return { ok: false, reason: 'there is no edit to undo in this block' };

    let idx;
    if (selector === 0) {
      idx = -1;
      for (let i = n - 1; i >= 0; i -= 1) if (this.#edits[i].live) { idx = i; break; }
      if (idx < 0) return { ok: false, reason: 'every edit in this block is already undone' };
    } else if (selector > 0) {
      idx = selector - 1;
    } else {
      idx = n + selector;
    }
    if (idx < 0 || idx >= n) {
      return { ok: false, reason: `edit ${selector} is outside this block's ${n} edit(s)` };
    }
    if (!this.#edits[idx].live) {
      return { ok: false, reason: `edit ${idx + 1} is already undone` };
    }
    this.#edits[idx].live = false;
    this.#removed.push(idx);
    return { ok: true };
  }

  undoByText(text) {
    let found = -1;
    let matches = 0;
    for (let i = this.#edits.length - 1; i >= 0; i -= 1) {
      if (this.#edits[i].text !== text) continue;
      matches += 1;
      if (found < 0 && this.#edits[i].live) found = i;
    }
    if (found < 0) {
      return {
        ok: false,
        ambiguous: false,
        reason: matches > 0 ? 'that edit is already undone' : `no edit in this block matches '${text}'`,
      };
    }
    this.#edits[found].live = false;
    this.#removed.push(found);
    return { ok: true, ambiguous: matches > 1 };
  }

  redo(times) {
    if (this.#removed.length === 0) return { ok: false, reason: 'nothing to redo in this block' };
    for (let k = 0; k < times && this.#removed.length > 0; k += 1) {
      this.#edits[this.#removed.pop()].live = true;
    }
    return { ok: true };
  }

  // Emits the undo{steps} + replay that makes the applied state equal the live set.
  reconcile(out, block, line, col) {
    const live = [];
    for (let i = 0; i < this.#edits.length; i += 1) if (this.#edits[i].live) live.push(i);

    let k = 0;
    while (k < this.#applied.length && k < live.length && this.#applied[k] === live[k]) k += 1;

    const steps = this.#applied.length - k;
    if (steps > 0) {
      out.push({ kind: 'undo', block, line, col, len: 0, editIndex: 0, strs: [], toks: [], nums: [steps] });
    }
    for (let i = k; i < live.length; i += 1) out.push(this.#edits[live[i]].op);
    this.#applied = live;
  }

  // A new @frame starts a fresh base: nothing from before is undoable.
  reset() {
    this.#edits = [];
    this.#applied = [];
    this.#removed = [];
  }
}

// Built exactly like scriptLower's normalizedText, so a selector and the edit it names
// produce the same string.
const selectorText = (st) => {
  let out = st.args[0].text.slice(1).toLowerCase();
  for (let i = 1; i < st.args.length; i += 1) {
    if (st.args[i].kind === 'comment') continue;
    out += ` ${st.args[i].text.toLowerCase()}`;
  }
  return out;
};

export const applyHistoryStmt = (st, isRedo, ledger, diags) => {
  if (isRedo) {
    let times = 1;
    if (st.args.length > 0 && st.args[0].kind === 'number') times = parseInt(st.args[0].text, 10);
    if (!Number.isFinite(times) || times < 1) times = 1;
    const r = ledger.redo(times);
    if (!r.ok) {
      diags.push(makeDiag('warning', 'W_NOTHING_TO_REDO', tokenOfStmt(st), r.reason));
      return false;
    }
    return true;
  }

  if (st.args.length === 0) {
    const r = ledger.undoIndex(0);
    if (!r.ok) {
      diags.push(makeDiag('error', 'E_UNDO_NO_EDITS', tokenOfStmt(st), r.reason));
      return false;
    }
    return true;
  }

  if (st.args[0].kind === 'directive') {
    const r = ledger.undoByText(selectorText(st));
    if (!r.ok) {
      diags.push(makeDiag('error', 'E_UNDO_OUT_OF_RANGE', tokenOfStmt(st), r.reason));
      return false;
    }
    if (r.ambiguous) {
      diags.push(makeDiag('warning', 'W_AMBIGUOUS_UNDO', tokenOfStmt(st),
        'several edits match — the last one was undone'));
    }
    return true;
  }

  let any = false;
  for (const t of st.args) {
    if (t.kind === 'punct') continue;
    if (t.kind !== 'number') {
      diags.push(makeDiag('error', 'E_BAD_TOKEN', t, `'${t.text}' is not an edit number`));
      return false;
    }
    const selector = parseInt(t.text, 10);
    if (selector === 0) {
      diags.push(makeDiag('error', 'E_UNDO_OUT_OF_RANGE', t,
        'edits are numbered from 1; use -1 for the last one'));
      return false;
    }
    const r = ledger.undoIndex(selector);
    if (!r.ok) {
      diags.push(makeDiag('error', ledger.editCount === 0 ? 'E_UNDO_NO_EDITS' : 'E_UNDO_OUT_OF_RANGE',
        t, r.reason));
      return false;
    }
    any = true;
  }
  if (!any) {
    const r = ledger.undoIndex(0);
    if (!r.ok) {
      diags.push(makeDiag('error', 'E_UNDO_NO_EDITS', tokenOfStmt(st), r.reason));
      return false;
    }
  }
  return true;
};

File: core/script/scriptUndo.cpp
#include "scriptUndo.hpp"

#include <algorithm>
#include <utility>

namespace stencil::core::script {

  int EditLedger::addEdit(Op op, std::string text) {
    EditRec rec;
    rec.op = std::move(op);
    rec.text = std::move(text);
    edits.push_back(std::move(rec));
    applied.push_back(static_cast<int>(edits.size()) - 1);
    return static_cast<int>(edits.size());
  }

  bool EditLedger::undoIndex(int selector, std::string& reason) {
    const int n = static_cast<int>(edits.size());
    if (n == 0) {
      reason = "there is no edit to undo in this block";
      return false;
    }
    int idx;
    if (selector == 0) {  // bare @undo — the last edit still live
      idx = -1;
      for (int i = n - 1; i >= 0; --i)
        if (edits[static_cast<std::size_t>(i)].live) { idx = i; break; }
      if (idx < 0) {
        reason = "every edit in this block is already undone";
        return false;
      }
    } else if (selector > 0) {
      idx = selector - 1;
    } else {
      idx = n + selector;  // -1 is the last edit
    }
    if (idx < 0 || idx >= n) {
      reason = "edit " + std::to_string(selector) + " is outside this block's " +
               std::to_string(n) + " edit(s)";
      return false;
    }
    EditRec& rec = edits[static_cast<std::size_t>(idx)];
    if (!rec.live) {
      reason = "edit " + std::to_string(idx + 1) + " is already undone";
      return false;
    }
    rec.live = false;
    removed.push_back(idx);
    return true;
  }

  bool EditLedger::undoByText(const std::string& text, bool& ambiguous, std::string& reason) {
    ambiguous = false;
    int found = -1, matches = 0;
    for (int i = static_cast<int>(edits.size()) - 1; i >= 0; --i) {
      if (edits[static_cast<std::size_t>(i)].text != text) continue;
      ++matches;
      if (found < 0 && edits[static_cast<std::size_t>(i)].live) found = i;
    }
    if (found < 0) {
      reason = matches > 0 ? "that edit is already undone"
                           : "no edit in this block matches '" + text + "'";
      return false;
    }
    ambiguous = matches > 1;
    edits[static_cast<std::size_t>(found)].live = false;
    removed.push_back(found);
    return true;
  }

  bool EditLedger::redo(int times, std::string& reason) {
    if (removed.empty()) {
      reason = "nothing to redo in this block";
      return false;
    }
    for (int k = 0; k < times && !removed.empty(); ++k) {
      edits[static_cast<std::size_t>(removed.back())].live = true;
      removed.pop_back();
    }
    return true;
  }

  bool EditLedger::reconcile(std::vector<Op>& out, int block, int line, int col) {
    std::vector<int> live;
    for (std::size_t i = 0; i < edits.size(); ++i)
      if (edits[i].live) live.push_back(static_cast<int>(i));

    std::size_t k = 0;
    while (k < applied.size() && k < live.size() && applied[k] == live[k]) ++k;

    const std::size_t steps = applied.size() - k;
    const std::size_t appended = (steps > 0 ? 1u : 0u) + (live.size() - k);
    if (appended > 0 && out.size() + appended >= static_cast<std::size_t>(MAX_OPS))
      return false;

    if (steps > 0) {
      Op undo;
      undo.kind = OpKind::UNDO;
      undo.block = block;
      undo.line = line;
      undo.col = col;
      undo.nums.assign(1, static_cast<double>(steps));
      out.push_back(std::move(undo));
    }
    for (std::size_t i = k; i < live.size(); ++i)
      out.push_back(edits[static_cast<std::size_t>(live[i])].op);
    applied = std::move(live);
    return true;
  }

  void EditLedger::reset() {
    edits.clear();
    applied.clear();
    removed.clear();
  }

}  // namespace stencil::core::script
